File: server/internal/models/user_device.go
package models

import (
	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type UserDevice struct {
	UserID    uuid.UUID
	Device    Device
	CreatedAt int64
	UpdatedAt int64
}

func UserDeviceFromDB(row *storage.DivaUserDevice, device *Device) *UserDevice {
	return &UserDevice{
		UserID:    row.UserID,
		Device:    *device,
		CreatedAt: row.CreatedAt,
		UpdatedAt: row.UpdatedAt,
	}
}

func (ud *UserDevice) Response() *responses.UserDeviceResponse {
	return &responses.UserDeviceResponse{
		UserID:     ud.UserID.String(),
		DeviceID:   ud.Device.ID.String(),
		DeviceName: ud.Device.Name,
		CreatedAt:  ud.CreatedAt,
		UpdatedAt:  ud.UpdatedAt,
	}
}

File: client/packages/shared/app-lib/diva-lib-models-core/src/commonMain/kotlin/io/github/juevigrace/diva/lib/models/user/device/UserDevice.kt
package io.github.juevigrace.diva.lib.models.user.device

import io.github.juevigrace.diva.lib.models.api.user.device.UserDeviceResponse
import io.github.juevigrace.diva.lib.models.device.Device
import kotlin.time.Instant
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
data class UserDevice(
    val userId: Uuid,
    val device: Device,
    val createdAt: Instant,
    val updatedAt: Instant,
) {
    companion object {
        fun fromResponse(response: UserDeviceResponse): UserDevice {
            return UserDevice(
                userId = Uuid.parse(response.userId),
                device = Device(
                    id = Uuid.parse(response.deviceId),
                    name = response.deviceName,
                    createdAt = Instant.fromEpochMilliseconds(response.createdAt),
                    updatedAt = Instant.fromEpochMilliseconds(response.updatedAt),
                ),
                createdAt = Instant.fromEpochMilliseconds(response.createdAt),
                updatedAt = Instant.fromEpochMilliseconds(response.updatedAt),
            )
        }
    }
}
